micro-rdk/src/common/system.rs

```diff
@@ -106,37 +130,88 @@ pub(crate) async fn force_shutdown(app_client: Option<AppClient>) {
-        Some(SystemEvent::DeepSleep(dur)) => {
+        Some(SystemEvent::DeepSleep {
+            duration,
+            ulp_enabled,
+        }) => {
             #[cfg(feature = "esp32")]
             {
-                if let Some(dur) = dur {
+                let mut result: sys::esp_err_t;
+
+                // disable other wakeup sources before setting
+                unsafe {
+                    result = sys::esp_sleep_disable_wakeup_source(
+                        sys::esp_sleep_source_t_ESP_SLEEP_WAKEUP_ALL,
+                    );
+                }
+
+                if result != sys::ESP_OK {
+                    log::error!("failed to clear wakeup sources before setting: {}", result);
+                }
+
+                if ulp_enabled {
+                    log::info!("enabling ULP wakeup");
+
+                    unsafe {
+                        result = sys::esp_sleep_enable_ulp_wakeup();
+                    }
+
+                    match result {
+                        sys::ESP_OK => {
+                            log::info!("ULP wakeup enabled");
+                            // Due to the many states the main device and ULP co-processor can possibly be in,
+                            // we prioritize ULP-based wakeup and forgo other options.
+                            if duration.is_some() {
+                                log::warn!(
+                                    "Duration-based wakeup will be ignored in favor of ULP wakeup"
+                                );
+                            }
+                            log::warn!("Esp32 entering deep sleep indefinitely until ULP wakeup");
+                            unsafe {
+                                sys::esp_deep_sleep_start();
+                            }
+                        }
+                        sys::ESP_ERR_NOT_SUPPORTED => {
+                            log::error!("additional current by touch enabled");
+                        }
+                        sys::ESP_ERR_INVALID_STATE => {
+                            log::error!("co-processor not enabled or wakeup trigger conflicts with ulp wakeup");
+                        }
+                        _ => log::error!("failed to enable ULP: {:?}", result),
+                    }
+                }
+
+                if let Some(dur) = duration {
                     let dur_micros = dur.as_micros() as u64;
-                    let result: crate::esp32::esp_idf_svc::sys::esp_err_t;
                     unsafe {
-                        result = crate::esp32::esp_idf_svc::sys::esp_sleep_enable_timer_wakeup(
-                            dur_micros,
-                        );
+                        result = sys::esp_sleep_enable_timer_wakeup(dur_micros);
                     }
-                    if result != crate::esp32::esp_idf_svc::sys::ESP_OK {
+                    if result != sys::ESP_OK {
                         unreachable!("duration requested too long")
```

Comment:
we should take advantage of just logging the error and go to sleep anyway (at least for now)